### PR TITLE
SM4 optimization for ARM by HW instruction

### PR DESCRIPTION
**Pull request by daniel-hu-arm**
This patch implements the SM4 optimization for ARM processor,
using SM4 HW instruction, which is an optional feature of
crypto extension for aarch64 V8.

Tested on modern ARM micro-architecture with SM4 support, the performance
 uplift can be observed around 8X ~ 40X over existing C implementation
 in openssl. Algorithms that can be parallelized (like CTR, ECB, CBC
 decryption) are on higher end, with algorithm like CBC encryption on
 lower end (due to its intra-block dependency)

Signed-off-by: Daniel Hu <Daniel.Hu@arm.com>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by uudiin** [comment]
Great job, thanks for your patch, we did some performance benchmark testing,
The following is the performance data:

Benchmark on T-Head Yitian-710 2.75GHz:

  Before:
  type      16 bytes     64 bytes    256 bytes    1024 bytes   8192 bytes  16384 bytes
  SM4-CTR  105787.80k   107837.87k   108380.84k   108462.08k   108549.46k   108554.92k
  SM4-ECB  111924.58k   118173.76k   119776.00k   120093.70k   120264.02k   120274.94k
  SM4-CBC  106428.09k   109190.98k   109674.33k   109774.51k   109827.41k   109827.41k

  After (7.4x - 36.6x faster):
  type      16 bytes     64 bytes    256 bytes    1024 bytes   8192 bytes  16384 bytes
  SM4-CTR  781979.02k  2432994.28k  3437753.86k  3834177.88k  3963715.58k  3974556.33k
  SM4-ECB  937590.69k  2941689.02k  3945751.81k  4328655.87k  4459181.40k  4468692.31k
  SM4-CBC  890639.88k  1027746.58k  1050621.78k  1056696.66k  1058613.93k  1058701.31k

Tested-by: Tianjia Zhang \<tianjia.zhang@linux.alibaba.com\>

**Comment by mattcaswell** [comment]
I'd like to see someone review the assembler code (does not need to be an OpenSSL committer, just someone familiar with aarch64 assembler).

**Comment by mattcaswell** [comment]
> I'd like to see someone review the assembler code

Maybe @zorrorffm?

**Comment by zorrorffm** [comment]
I've reviewed the assembly code inside arm. It looks good to me.

File: crypto/evp/e_sm4.c
Comment (paulidale) [inline crypto/evp/e_sm4.c:136]: style nit: blank line separating declarations from code.

File: crypto/evp/e_sm4.c
Comment (paulidale) [inline crypto/evp/e_sm4.c:153]: `if (dat->stream.ecb != NULL)`

File: crypto/evp/e_sm4.c
Comment (paulidale) [inline crypto/evp/e_sm4.c:156]: style nit: `for` on the next line please.

File: crypto/evp/e_sm4.c
Comment (paulidale) [inline crypto/evp/e_sm4.c:167]: style nit: blank line above.

File: crypto/evp/e_sm4.c
Comment (paulidale) [inline crypto/evp/e_sm4.c:119]: style nit: blank line separating declarations from code.

**Review by paulidale — CHANGES_REQUESTED** [review]
Just some style nits.

Nice work.

File: crypto/evp/e_sm4.c
Comment (daniel-hu-arm) [reply crypto/evp/e_sm4.c:119]: interesting, do not see this blank line in code using vi, perhaps this is just how review tool show the code side-by-side?

File: crypto/evp/e_sm4.c
Comment (daniel-hu-arm) [reply crypto/evp/e_sm4.c:136]: ditto

File: crypto/evp/e_sm4.c
Comment (daniel-hu-arm) [reply crypto/evp/e_sm4.c:167]: ditto

File: crypto/evp/e_sm4.c
Comment (daniel-hu-arm) [reply crypto/evp/e_sm4.c:119]: ok, just to ensure i understand the problem correctly

instead of:
static int sm4_cbc_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
                          const unsigned char *in, size_t len)
{
    EVP_SM4_KEY *dat = EVP_C_DATA(EVP_SM4_KEY,ctx);

should we write:
static int sm4_cbc_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
                          const unsigned char *in, size_t len)
<a blank line here>
{
    EVP_SM4_KEY *dat = EVP_C_DATA(EVP_SM4_KEY,ctx);

Comment (t8m) [reply crypto/evp/e_sm4.c:120]: The blank line should be placed after the local variable declarations. Not before the `{`

File: crypto/evp/e_sm4.c
Comment (t8m) [inline crypto/evp/e_sm4.c:117]: The blank line should not be here.

File: crypto/evp/e_sm4.c
Comment (t8m) [reply crypto/evp/e_sm4.c:120]: So:
```
    EVP_SM4_KEY *dat = EVP_C_DATA(EVP_SM4_KEY,ctx);

    if (dat->stream.cbc)
```

File: crypto/evp/e_sm4.c
Comment (daniel-hu-arm) [reply crypto/evp/e_sm4.c:117]: now i get it! thanks.

**Review by t8m — APPROVED** [review]

**Comment by uudiin** [comment]
@daniel-hu-arm Hi, Is it possible to put Yitian's performance data on commit message? This data can now be made public. Thanks.

**Comment by paulidale** [comment]
Agreed, let's include the performance data in the commit please.

**Comment by daniel-hu-arm** [comment]
@paulidale new commit pushed to include the Yitian perf data

**Comment by t8m** [comment]
The force push just changed the commit message.

**Comment by t8m** [comment]
OK after merging SM3 support this needs to be rebased. Could you please do that? @daniel-hu-arm

**Comment by daniel-hu-arm** [comment]
@t8m working on it.

**Comment by daniel-hu-arm** [comment]
@t8m rebased to upstream master. (had to do a "force" push back to fork repo, still not sure why)

**Comment by t8m** [comment]
@paulidale still OK?

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged a while ago. Thank you for your contribution.

**Comment by zzl360** [comment]
it seems that https://github.com/openssl/openssl/blob/d2d2401aed7ff45f4c013201944e1218dce12da7/providers/implementations/ciphers/cipher_sm4_gcm_hw.c#L29 doesn't work?
I don't find any code invoke this ctr function.
maybe we should write new function to change https://github.com/openssl/openssl/blob/d2d2401aed7ff45f4c013201944e1218dce12da7/providers/implementations/ciphers/cipher_sm4_gcm_hw.c#L49 like https://github.com/openssl/openssl/blob/d2d2401aed7ff45f4c013201944e1218dce12da7/providers/implementations/ciphers/cipher_aes_gcm_hw.c#L131?

**Comment by t8m** [comment]
Perhaps you could just add the non-NULL ctr case to the ossl_gcm_cipher_update() instead of writing a completely new function?

**Comment by daniel-hu-arm** [comment]
the scope of the SM4 patch was mainly to focus on hw acceleration and evp interface. but anyone can call these hw accelerated routine we implemented from high-level code. we will look into it and see if we can do more about it.

**Comment by daniel-hu-arm** [comment]
we have a patch to enable HW acceleration for GCM on ARM now, will be up-streamed soon.

**Comment by zzl360** [comment]
I think it may enable HW acceleration for CCM on ARM,just add HW acceleration code in providers/implementations/ciphers/cipher_sm4_ccm_hw.c

**Comment by zzl360** [comment]
on the cbc-decrypt mode, the benchmark could be up to 4189.2 MB/s when 16384 bytes

**Pull request closed without merging**
